{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train a CNN regression to predict the mCG level of all DMRs\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time, re\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quantify mCG at the enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bed_file='enhancers_reptile.bed'\n",
    "# allc_dir='/cndd2/fangming/scf_enhancers/data/bulk/round2/mc/'\n",
    "# allc_files=glob.glob(allc_dir+'Round2*CGN-Merge.allc.tsv.gz')\n",
    "\n",
    "# # bed_file='foo.bed'\n",
    "# # allc_files = [allc_files[0]]\n",
    "\n",
    "# output_files = ['enh.'+os.path.split(a)[1][:-22]+'.bed.gz' for a in allc_files]\n",
    "# cap = 0 # no counts cap\n",
    "# nprocs = 8\n",
    "# compress = True\n",
    "# bed_file_name_column = False\n",
    "# chr_prefix = True\n",
    "# contexts = ['CG']\n",
    "\n",
    "# run_mc_region_level(allc_files, \n",
    "#                     output_files, \n",
    "#                     bed_file, \n",
    "#                     chr_prefix=chr_prefix,\n",
    "#                     bed_file_name_column=bed_file_name_column,\n",
    "#                     contexts=contexts,\n",
    "#                     compress=compress, \n",
    "#                     cap=cap,\n",
    "#                     nprocs=nprocs);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Gather data from all the individual bed files\n",
    "# df = pd.DataFrame();\n",
    "# for fn in output_files:\n",
    "    \n",
    "#     _, sample = os.path.split(fn)\n",
    "#     sample = re.sub('(enh\\.|\\.bed)','',sample)\n",
    "\n",
    "#     x = pd.read_csv(fn+'.gz',sep='\\t')\n",
    "#     x=x.rename(mapper={'mCG': sample+'_'+'mCG', 'CG': sample+'_'+'CG'}, axis=1)\n",
    "#     if df.shape[0]<1:\n",
    "#         df = x\n",
    "#     else:\n",
    "#         df = df.join(x.iloc[:,[3,4]])\n",
    "        \n",
    "# df['enh_pos'] = 'chr'+df['chr'].astype(str)+'_'+df['start'].astype(str)+'_'+df['end'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.to_csv('enh.AllClusters.mCG.bed.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for fn in output_files:\n",
    "#     os.remove(fn)\n",
    "#     print('Removed file %s' % fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load enhancer data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map the genes to closest enhancers\n",
    "datadir = '/cndd/emukamel/enhancer_modeling/enhancer_data/'\n",
    "genes2enh = pd.read_csv(datadir+'genes.gencode.vM16.closest_enhancers.bed',sep='\\t',\n",
    "                        names = ['gene_chr','gene_start','gene_end','ensid','gene_name','enh_chr','enh_start','enh_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes2enh['enh_pos'] = genes2enh['enh_chr']+'_'+genes2enh['enh_start'].astype(str)+'_'+genes2enh['enh_end'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes2enh.ensid = [re.sub(\"\\..*\",\"\",x) for x in genes2enh.ensid]\n",
    "genes2enh = genes2enh.set_index('ensid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ensid</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102693</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064842</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>Gm26206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102851</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3252757</td>\n",
       "      <td>3253236</td>\n",
       "      <td>ENSMUSG00000102851</td>\n",
       "      <td>Gm18956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000103377</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3365731</td>\n",
       "      <td>3368549</td>\n",
       "      <td>ENSMUSG00000103377</td>\n",
       "      <td>Gm37180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000095134</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90753057</td>\n",
       "      <td>90763485</td>\n",
       "      <td>ENSMUSG00000095134</td>\n",
       "      <td>Mid1-ps1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000095366</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90754513</td>\n",
       "      <td>90754821</td>\n",
       "      <td>ENSMUSG00000095366</td>\n",
       "      <td>Gm21860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000096768</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90784738</td>\n",
       "      <td>90816464</td>\n",
       "      <td>ENSMUSG00000096768</td>\n",
       "      <td>Erdr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000099871</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90837413</td>\n",
       "      <td>90844040</td>\n",
       "      <td>ENSMUSG00000099871</td>\n",
       "      <td>Gm21742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000096850</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90838869</td>\n",
       "      <td>90839177</td>\n",
       "      <td>ENSMUSG00000096850</td>\n",
       "      <td>Gm21748</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53379 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     chr     start       end               ensid  \\\n",
       "ensid                                                              \n",
       "ENSMUSG00000102693  chr1   3073253   3074322  ENSMUSG00000102693   \n",
       "ENSMUSG00000064842  chr1   3102016   3102125  ENSMUSG00000064842   \n",
       "ENSMUSG00000051951  chr1   3205901   3671498  ENSMUSG00000051951   \n",
       "ENSMUSG00000102851  chr1   3252757   3253236  ENSMUSG00000102851   \n",
       "ENSMUSG00000103377  chr1   3365731   3368549  ENSMUSG00000103377   \n",
       "...                  ...       ...       ...                 ...   \n",
       "ENSMUSG00000095134  chrY  90753057  90763485  ENSMUSG00000095134   \n",
       "ENSMUSG00000095366  chrY  90754513  90754821  ENSMUSG00000095366   \n",
       "ENSMUSG00000096768  chrY  90784738  90816464  ENSMUSG00000096768   \n",
       "ENSMUSG00000099871  chrY  90837413  90844040  ENSMUSG00000099871   \n",
       "ENSMUSG00000096850  chrY  90838869  90839177  ENSMUSG00000096850   \n",
       "\n",
       "                             gene  \n",
       "ensid                              \n",
       "ENSMUSG00000102693  4933401J01Rik  \n",
       "ENSMUSG00000064842        Gm26206  \n",
       "ENSMUSG00000051951           Xkr4  \n",
       "ENSMUSG00000102851        Gm18956  \n",
       "ENSMUSG00000103377        Gm37180  \n",
       "...                           ...  \n",
       "ENSMUSG00000095134       Mid1-ps1  \n",
       "ENSMUSG00000095366        Gm21860  \n",
       "ENSMUSG00000096768          Erdr1  \n",
       "ENSMUSG00000099871        Gm21742  \n",
       "ENSMUSG00000096850        Gm21748  \n",
       "\n",
       "[53379 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv(datadir+'genes.gencode.vM16.bed',sep='\\t',header=None,\n",
    "                   names=['chr','start','end','ensid','gene'])\n",
    "genes['ensid'] = [re.sub('\\..*','',ensid) for ensid in genes['ensid']]\n",
    "genes = genes.set_index('ensid',drop=False)\n",
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load RNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "datadir='/cndd2/ethan/projects/scf_paper/modeling/working_data/reptile/data_for_model_01152020/' # 10x vs scRNA\n",
    "rna = pd.read_csv(datadir+'rna_for_y_values.tsv',sep='\\t')\n",
    "rcol = rna.columns.values\n",
    "rcol[1] = 'ensid'\n",
    "rna.columns = rcol\n",
    "# rna=rna.drop('10-1',axis=1)\n",
    "# rna_data = rna.iloc[:,2:].to_numpy()\n",
    "\n",
    "# Variable genes\n",
    "hvg = np.var(rna.iloc[:,2:].to_numpy(),axis=1);\n",
    "hvg = hvg> np.percentile(hvg,50)\n",
    "np.sum(hvg)\n",
    "\n",
    "rna = rna[hvg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = rna.set_index('ensid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset to the HVGs\n",
    "genes2enhu = genes2enh.loc[rna.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(57,)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1-1</th>\n",
       "      <th>1-2</th>\n",
       "      <th>10-1</th>\n",
       "      <th>2-1</th>\n",
       "      <th>2-2</th>\n",
       "      <th>2-3</th>\n",
       "      <th>3-1</th>\n",
       "      <th>3-2</th>\n",
       "      <th>3-3</th>\n",
       "      <th>4-1</th>\n",
       "      <th>...</th>\n",
       "      <th>6-4</th>\n",
       "      <th>6-5</th>\n",
       "      <th>6-6</th>\n",
       "      <th>7-1</th>\n",
       "      <th>7-2</th>\n",
       "      <th>8-1</th>\n",
       "      <th>8-2</th>\n",
       "      <th>8-3</th>\n",
       "      <th>9-1</th>\n",
       "      <th>9-2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>54331</td>\n",
       "      <td>34776</td>\n",
       "      <td>2849</td>\n",
       "      <td>105663</td>\n",
       "      <td>60324</td>\n",
       "      <td>16182</td>\n",
       "      <td>86593</td>\n",
       "      <td>66870</td>\n",
       "      <td>196</td>\n",
       "      <td>56461</td>\n",
       "      <td>...</td>\n",
       "      <td>9949</td>\n",
       "      <td>3968</td>\n",
       "      <td>700</td>\n",
       "      <td>5352</td>\n",
       "      <td>13660</td>\n",
       "      <td>10301</td>\n",
       "      <td>8077</td>\n",
       "      <td>816</td>\n",
       "      <td>2906</td>\n",
       "      <td>2122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000089699</th>\n",
       "      <td>4431</td>\n",
       "      <td>2933</td>\n",
       "      <td>287</td>\n",
       "      <td>8416</td>\n",
       "      <td>4705</td>\n",
       "      <td>1312</td>\n",
       "      <td>7140</td>\n",
       "      <td>5589</td>\n",
       "      <td>18</td>\n",
       "      <td>5180</td>\n",
       "      <td>...</td>\n",
       "      <td>880</td>\n",
       "      <td>350</td>\n",
       "      <td>77</td>\n",
       "      <td>449</td>\n",
       "      <td>1299</td>\n",
       "      <td>718</td>\n",
       "      <td>523</td>\n",
       "      <td>48</td>\n",
       "      <td>206</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102343</th>\n",
       "      <td>73</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>40</td>\n",
       "      <td>9</td>\n",
       "      <td>71</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025900</th>\n",
       "      <td>227</td>\n",
       "      <td>104</td>\n",
       "      <td>6</td>\n",
       "      <td>249</td>\n",
       "      <td>110</td>\n",
       "      <td>71</td>\n",
       "      <td>245</td>\n",
       "      <td>248</td>\n",
       "      <td>1</td>\n",
       "      <td>324</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025902</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000097934</th>\n",
       "      <td>39</td>\n",
       "      <td>17</td>\n",
       "      <td>8</td>\n",
       "      <td>41</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025931</th>\n",
       "      <td>1208</td>\n",
       "      <td>529</td>\n",
       "      <td>13</td>\n",
       "      <td>1566</td>\n",
       "      <td>804</td>\n",
       "      <td>642</td>\n",
       "      <td>1505</td>\n",
       "      <td>1266</td>\n",
       "      <td>5</td>\n",
       "      <td>1229</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>51</td>\n",
       "      <td>5</td>\n",
       "      <td>76</td>\n",
       "      <td>113</td>\n",
       "      <td>47</td>\n",
       "      <td>80</td>\n",
       "      <td>103</td>\n",
       "      <td>208</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000041809</th>\n",
       "      <td>100</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>134</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000041779</th>\n",
       "      <td>134</td>\n",
       "      <td>79</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>65</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000099971</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      1-1    1-2  10-1     2-1    2-2    2-3    3-1    3-2  \\\n",
       "ENSMUSG00000051951  54331  34776  2849  105663  60324  16182  86593  66870   \n",
       "ENSMUSG00000089699   4431   2933   287    8416   4705   1312   7140   5589   \n",
       "ENSMUSG00000102343     73     15     0      56     40      9     71     44   \n",
       "ENSMUSG00000025900    227    104     6     249    110     71    245    248   \n",
       "ENSMUSG00000025902      1      0     0      10      0      0      0      0   \n",
       "...                   ...    ...   ...     ...    ...    ...    ...    ...   \n",
       "ENSMUSG00000097934     39     17     8      41     27     19     58     50   \n",
       "ENSMUSG00000025931   1208    529    13    1566    804    642   1505   1266   \n",
       "ENSMUSG00000041809    100     46     0      73     30     22    134    152   \n",
       "ENSMUSG00000041779    134     79     1      80     24     22     65     70   \n",
       "ENSMUSG00000099971      0      0     0       0      0      0      0      2   \n",
       "\n",
       "                    3-3    4-1  ...   6-4   6-5  6-6   7-1    7-2    8-1  \\\n",
       "ENSMUSG00000051951  196  56461  ...  9949  3968  700  5352  13660  10301   \n",
       "ENSMUSG00000089699   18   5180  ...   880   350   77   449   1299    718   \n",
       "ENSMUSG00000102343    0     66  ...     3     2    0     0      7      6   \n",
       "ENSMUSG00000025900    1    324  ...    12    13    1    17     21     20   \n",
       "ENSMUSG00000025902    0      2  ...     0     0    0     0      0      0   \n",
       "...                 ...    ...  ...   ...   ...  ...   ...    ...    ...   \n",
       "ENSMUSG00000097934    0     70  ...     0     1    0     6      7      1   \n",
       "ENSMUSG00000025931    5   1229  ...    20    51    5    76    113     47   \n",
       "ENSMUSG00000041809    0    136  ...     3     2    0     5     13     15   \n",
       "ENSMUSG00000041779    1     99  ...     9     2    0     4      8     14   \n",
       "ENSMUSG00000099971    0      1  ...     0     0    0     0      0      0   \n",
       "\n",
       "                     8-2  8-3   9-1   9-2  \n",
       "ENSMUSG00000051951  8077  816  2906  2122  \n",
       "ENSMUSG00000089699   523   48   206   182  \n",
       "ENSMUSG00000102343     1    1     6    13  \n",
       "ENSMUSG00000025900     4    0    20    20  \n",
       "ENSMUSG00000025902     0    0     0     0  \n",
       "...                  ...  ...   ...   ...  \n",
       "ENSMUSG00000097934     0    0     3     7  \n",
       "ENSMUSG00000025931    80  103   208   308  \n",
       "ENSMUSG00000041809     0    1    20    16  \n",
       "ENSMUSG00000041779     3    0     3     6  \n",
       "ENSMUSG00000099971     0    0     0     0  \n",
       "\n",
       "[100 rows x 29 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load different RNA datasets\n",
    "rnadir = '/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/'\n",
    "rna_modalities = ['Table_S5a_gene_counts_scRNA_SMART.tsv.gz',\n",
    "    'Table_S5b_gene_counts_scRNA_10X_v3_A.tsv.gz',\n",
    "    'Table_S5c_gene_counts_scRNA_10X_v2_A.tsv.gz',\n",
    "    'Table_S5d_gene_counts_snRNA_SMART.tsv.gz',\n",
    "    'Table_S5e_gene_counts_snRNA_10X_v3_B.tsv.gz',\n",
    "    'Table_S5f_gene_counts_snRNA_10X_v3_A.tsv.gz',\n",
    "#     'Table_S5g_gene_counts_open_chromatin.tsv.gz',\n",
    "#     'Table_S5h_gene_counts_DNA_methylation.xlsx.gz'\n",
    "    ]\n",
    "\n",
    "all_rna = {'modality':[], 'data':[], 'celltypes':['1-1-1', '1-2-1', '1-2-2', '10-1-1', '2-1-1', '2-1-2',\n",
    "       '2-2-1', '2-2-2', '2-2-3', '2-2-4', '2-3-1', '2-3-2', '2-3-3', '3-1-1',\n",
    "       '3-1-2', '3-2-1', '3-2-2', '3-3-1', '4-1-1', '4-1-2', '4-2-1', '4-2-2',\n",
    "       '4-2-3', '5-1-1', '5-1-2', '5-2-1', '5-2-2', '5-2-3', '5-3-1', '5-3-2',\n",
    "       '5-4-1', '5-5-1', '6-1-1', '6-1-2', '6-2-1', '6-3-1', '6-3-2', '6-3-3',\n",
    "       '6-4-1', '6-4-2', '6-5-1', '6-5-2', '6-5-3', '6-6-1', '7-1-1', '7-1-2',\n",
    "       '7-2-1', '7-2-2', '8-1-1', '8-1-2', '8-2-1', '8-3-1', '8-3-2', '9-1-1',\n",
    "       '9-2-1', '9-2-2']}\n",
    "scf_l2 = ['1-1-1', '1-2-1', '1-2-2', '10-1-1', '2-1-1', '2-1-2',\n",
    "       '2-2-1', '2-2-2', '2-2-3', '2-2-4', '2-3-1', '2-3-2', '2-3-3', '3-1-1',\n",
    "       '3-1-2', '3-2-1', '3-2-2', '3-3-1', '4-1-1', '4-1-2', '4-2-1', '4-2-2',\n",
    "       '4-2-3', '5-1-1', '5-1-2', '5-2-1', '5-2-2', '5-2-3', '5-3-1', '5-3-2',\n",
    "       '5-4-1', '5-5-1', '6-1-1', '6-1-2', '6-2-1', '6-3-1', '6-3-2', '6-3-3',\n",
    "       '6-4-1', '6-4-2', '6-5-1', '6-5-2', '6-5-3', '6-6-1', '7-1-1', '7-1-2',\n",
    "       '7-2-1', '7-2-2', '8-1-1', '8-1-2', '8-2-1', '8-3-1', '8-3-2', '9-1-1',\n",
    "       '9-2-1', '9-2-2']\n",
    "scf_l1,scf_l1i = np.unique(np.array([re.sub('-[0-9]$','', i) for i in scf_l2]),return_inverse=True)\n",
    "scf_l1,scf_l1i\n",
    "\n",
    "rna_modalities = 'Table_S5b_gene_counts_scRNA_10X_v3_A.tsv.gz'\n",
    "fn = rnadir+ds\n",
    "data_l2 = pd.read_csv(fn,sep='\\t',nrows=100,index_col=0)\n",
    "data_l1 = data_l2.loc[:,[]]\n",
    "for i in range(len(scf_l1)):\n",
    "    data_l1.loc[:,scf_l1[i]] = data_l2.iloc[:,scf_l1i==i].sum(axis=1)\n",
    "data_l1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<function re.sub(pattern, repl, string, count=0, flags=0)>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/Table_S5a_gene_counts_scRNA_SMART.tsv.gz\n",
      "/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/Table_S5b_gene_counts_scRNA_10X_v3_A.tsv.gz\n",
      "/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/Table_S5c_gene_counts_scRNA_10X_v2_A.tsv.gz\n",
      "/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/Table_S5d_gene_counts_snRNA_SMART.tsv.gz\n",
      "/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/Table_S5e_gene_counts_snRNA_10X_v3_B.tsv.gz\n",
      "/cndd/emukamel/BICCN-Mouse-MOp/supp_tables/Table_S5_GeneByCluster/Table_S5f_gene_counts_snRNA_10X_v3_A.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "for ds in rna_modalities:\n",
    "    fn = rnadir+ds\n",
    "    print(fn)\n",
    "    data = pd.read_csv(fn,sep='\\t',nrows=100,index_col=0)\n",
    "#     print(data.head())\n",
    "\n",
    "    data = data.loc[:,all_rna['celltypes']]\n",
    "    data = genes2enhu.join(data)\n",
    "#     print(np.setdiff1d(data.index, genes2enhu.index).shape)\n",
    "#     data = data.loc[genes2enhu.index.values,:]\n",
    "    all_rna['modality'].append(fn)\n",
    "    all_rna['data'].append(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_chr', 'gene_start', 'gene_end', 'gene_name', 'enh_chr',\n",
       "       'enh_start', 'enh_end', 'enh_pos', '1-1-1', '1-2-1', '1-2-2', '10-1-1',\n",
       "       '2-1-1', '2-1-2', '2-2-1', '2-2-2', '2-2-3', '2-2-4', '2-3-1', '2-3-2',\n",
       "       '2-3-3', '3-1-1', '3-1-2', '3-2-1', '3-2-2', '3-3-1', '4-1-1', '4-1-2',\n",
       "       '4-2-1', '4-2-2', '4-2-3', '5-1-1', '5-1-2', '5-2-1', '5-2-2', '5-2-3',\n",
       "       '5-3-1', '5-3-2', '5-4-1', '5-5-1', '6-1-1', '6-1-2', '6-2-1', '6-3-1',\n",
       "       '6-3-2', '6-3-3', '6-4-1', '6-4-2', '6-5-1', '6-5-2', '6-5-3', '6-6-1',\n",
       "       '7-1-1', '7-1-2', '7-2-1', '7-2-2', '8-1-1', '8-1-2', '8-2-1', '8-3-1',\n",
       "       '8-3-2', '9-1-1', '9-2-1', '9-2-2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_rna['data'][1].columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load mCG data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cndd/emukamel/conda_envs/scanpy/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Round2.3-3_mCG</th>\n",
       "      <th>Round2.3-3_CG</th>\n",
       "      <th>Round2.8-2_mCG</th>\n",
       "      <th>Round2.8-2_CG</th>\n",
       "      <th>Round2.2-1_mCG</th>\n",
       "      <th>Round2.2-1_CG</th>\n",
       "      <th>Round2.6-1_mCG</th>\n",
       "      <th>...</th>\n",
       "      <th>Round2.5-5_CG</th>\n",
       "      <th>Round2.2-2_mCG</th>\n",
       "      <th>Round2.2-2_CG</th>\n",
       "      <th>Round2.6-4_mCG</th>\n",
       "      <th>Round2.6-4_CG</th>\n",
       "      <th>Round2.4-1_mCG</th>\n",
       "      <th>Round2.4-1_CG</th>\n",
       "      <th>Round2.6-2_mCG</th>\n",
       "      <th>Round2.6-2_CG</th>\n",
       "      <th>enh_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3083769</td>\n",
       "      <td>3083816</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>32</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1_3083769_3083816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3112234</td>\n",
       "      <td>3112467</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>27</td>\n",
       "      <td>28</td>\n",
       "      <td>304</td>\n",
       "      <td>315</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>174</td>\n",
       "      <td>177</td>\n",
       "      <td>15</td>\n",
       "      <td>67</td>\n",
       "      <td>149</td>\n",
       "      <td>164</td>\n",
       "      <td>37</td>\n",
       "      <td>118</td>\n",
       "      <td>chr1_3112234_3112467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3112770</td>\n",
       "      <td>3112986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>49</td>\n",
       "      <td>53</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>chr1_3112770_3112986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3113717</td>\n",
       "      <td>3114035</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>35</td>\n",
       "      <td>206</td>\n",
       "      <td>44</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>125</td>\n",
       "      <td>15</td>\n",
       "      <td>59</td>\n",
       "      <td>55</td>\n",
       "      <td>115</td>\n",
       "      <td>10</td>\n",
       "      <td>95</td>\n",
       "      <td>chr1_3113717_3114035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3118790</td>\n",
       "      <td>3118935</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>93</td>\n",
       "      <td>130</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>56</td>\n",
       "      <td>14</td>\n",
       "      <td>29</td>\n",
       "      <td>42</td>\n",
       "      <td>89</td>\n",
       "      <td>5</td>\n",
       "      <td>48</td>\n",
       "      <td>chr1_3118790_3118935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302091</th>\n",
       "      <td>X</td>\n",
       "      <td>169890644</td>\n",
       "      <td>169890829</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>62</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>30</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "      <td>chrX_169890644_169890829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302092</th>\n",
       "      <td>X</td>\n",
       "      <td>169918008</td>\n",
       "      <td>169918093</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_169918008_169918093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302093</th>\n",
       "      <td>X</td>\n",
       "      <td>169927115</td>\n",
       "      <td>169927682</td>\n",
       "      <td>6</td>\n",
       "      <td>35</td>\n",
       "      <td>75</td>\n",
       "      <td>79</td>\n",
       "      <td>243</td>\n",
       "      <td>709</td>\n",
       "      <td>44</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>306</td>\n",
       "      <td>412</td>\n",
       "      <td>89</td>\n",
       "      <td>94</td>\n",
       "      <td>204</td>\n",
       "      <td>417</td>\n",
       "      <td>318</td>\n",
       "      <td>340</td>\n",
       "      <td>chrX_169927115_169927682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302094</th>\n",
       "      <td>X</td>\n",
       "      <td>169935048</td>\n",
       "      <td>169935204</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>48</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>32</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>27</td>\n",
       "      <td>30</td>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>chrX_169935048_169935204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302095</th>\n",
       "      <td>X</td>\n",
       "      <td>169983814</td>\n",
       "      <td>169984232</td>\n",
       "      <td>37</td>\n",
       "      <td>82</td>\n",
       "      <td>124</td>\n",
       "      <td>178</td>\n",
       "      <td>760</td>\n",
       "      <td>1218</td>\n",
       "      <td>250</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>488</td>\n",
       "      <td>798</td>\n",
       "      <td>189</td>\n",
       "      <td>231</td>\n",
       "      <td>450</td>\n",
       "      <td>681</td>\n",
       "      <td>452</td>\n",
       "      <td>539</td>\n",
       "      <td>chrX_169983814_169984232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>302096 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr      start        end  Round2.3-3_mCG  Round2.3-3_CG  \\\n",
       "0        1    3083769    3083816               1              1   \n",
       "1        1    3112234    3112467              15             16   \n",
       "2        1    3112770    3112986               0              0   \n",
       "3        1    3113717    3114035               8             14   \n",
       "4        1    3118790    3118935              12             14   \n",
       "...     ..        ...        ...             ...            ...   \n",
       "302091   X  169890644  169890829               2              2   \n",
       "302092   X  169918008  169918093               2              2   \n",
       "302093   X  169927115  169927682               6             35   \n",
       "302094   X  169935048  169935204               3              3   \n",
       "302095   X  169983814  169984232              37             82   \n",
       "\n",
       "        Round2.8-2_mCG  Round2.8-2_CG  Round2.2-1_mCG  Round2.2-1_CG  \\\n",
       "0                    5              5              31             32   \n",
       "1                   27             28             304            315   \n",
       "2                    1              4              49             53   \n",
       "3                    9             16              35            206   \n",
       "4                    2              2              93            130   \n",
       "...                ...            ...             ...            ...   \n",
       "302091              10             10              62             70   \n",
       "302092               3              3              26             26   \n",
       "302093              75             79             243            709   \n",
       "302094               8              9              45             48   \n",
       "302095             124            178             760           1218   \n",
       "\n",
       "        Round2.6-1_mCG  ...  Round2.5-5_CG  Round2.2-2_mCG  Round2.2-2_CG  \\\n",
       "0                   13  ...              0              18             20   \n",
       "1                   57  ...              0             174            177   \n",
       "2                    7  ...              0              30             33   \n",
       "3                   44  ...              3              16            125   \n",
       "4                   21  ...              0              46             56   \n",
       "...                ...  ...            ...             ...            ...   \n",
       "302091               0  ...              0              39             40   \n",
       "302092               0  ...              0              14             15   \n",
       "302093              44  ...              5             306            412   \n",
       "302094              17  ...              0              31             32   \n",
       "302095             250  ...              0             488            798   \n",
       "\n",
       "        Round2.6-4_mCG  Round2.6-4_CG  Round2.4-1_mCG  Round2.4-1_CG  \\\n",
       "0                    2              3              15             16   \n",
       "1                   15             67             149            164   \n",
       "2                    1             10              24             33   \n",
       "3                   15             59              55            115   \n",
       "4                   14             29              42             89   \n",
       "...                ...            ...             ...            ...   \n",
       "302091               2             16              30             37   \n",
       "302092               0              5               6              6   \n",
       "302093              89             94             204            417   \n",
       "302094               7              7              27             30   \n",
       "302095             189            231             450            681   \n",
       "\n",
       "        Round2.6-2_mCG  Round2.6-2_CG                   enh_pos  \n",
       "0                   13             14      chr1_3083769_3083816  \n",
       "1                   37            118      chr1_3112234_3112467  \n",
       "2                    6             18      chr1_3112770_3112986  \n",
       "3                   10             95      chr1_3113717_3114035  \n",
       "4                    5             48      chr1_3118790_3118935  \n",
       "...                ...            ...                       ...  \n",
       "302091               1             43  chrX_169890644_169890829  \n",
       "302092               5              7  chrX_169918008_169918093  \n",
       "302093             318            340  chrX_169927115_169927682  \n",
       "302094              25             28  chrX_169935048_169935204  \n",
       "302095             452            539  chrX_169983814_169984232  \n",
       "\n",
       "[302096 rows x 62 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datadir = '/cndd/emukamel/enhancer_modeling/enhancer_data/'\n",
    "df = pd.read_csv(datadir+'enh.AllClusters.mCG.bed.gz')\n",
    "df['enh_pos'] = 'chr'+df['chr'].astype(str)+'_'+df['start'].astype(str)+'_'+df['end'].astype(str)\n",
    "df = df.drop(['Unnamed: 0'],axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cndd/emukamel/conda_envs/scanpy/lib/python3.7/site-packages/ipykernel_launcher.py:4: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  after removing the cwd from sys.path.\n",
      "/cndd/emukamel/conda_envs/scanpy/lib/python3.7/site-packages/ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in true_divide\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "# Compute mCH level\n",
    "cg = df.iloc[:,3:-1:2]\n",
    "mcg = df.iloc[:,4:-1:2]\n",
    "mlevel = mcg.to_numpy() / cg.to_numpy()\n",
    "hthresh = 2\n",
    "mlevel[ cg<hthresh ] = np.nan\n",
    "\n",
    "df_mlevel = pd.DataFrame(mlevel, columns = cg.columns, index=df.enh_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Round2.3-3_mCG     False\n",
       "Round2.8-2_mCG      True\n",
       "Round2.2-1_mCG      True\n",
       "Round2.6-1_mCG      True\n",
       "Round2.4-2_mCG      True\n",
       "Round2.1-1_mCG      True\n",
       "Round2.10-1_mCG    False\n",
       "Round2.3-2_mCG      True\n",
       "Round2.7-2_mCG      True\n",
       "Round2.5-1_mCG      True\n",
       "Round2.8-3_mCG     False\n",
       "Round2.6-6_mCG     False\n",
       "Round2.9-1_mCG     False\n",
       "Round2.3-1_mCG      True\n",
       "Round2.1-2_mCG      True\n",
       "Round2.5-2_mCG      True\n",
       "Round2.7-1_mCG      True\n",
       "Round2.5-4_mCG     False\n",
       "Round2.2-3_mCG      True\n",
       "Round2.9-2_mCG     False\n",
       "Round2.6-5_mCG      True\n",
       "Round2.6-3_mCG      True\n",
       "Round2.5-3_mCG      True\n",
       "Round2.8-1_mCG      True\n",
       "Round2.5-5_mCG     False\n",
       "Round2.2-2_mCG      True\n",
       "Round2.6-4_mCG      True\n",
       "Round2.4-1_mCG      True\n",
       "Round2.6-2_mCG      True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAD8CAYAAABw1c+bAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAEH1JREFUeJzt3X+QXWV9x/H3R6LtlNKCsiICMWojDjqCdht1ah2oSiFS0EptMv0BLTZqsa1TZ1paO+roH6V1qtMWRiYKA3YUqbbYdAhI6o9BOqIGGn4K8qNxSKQkgIL4ozb67R97Ml2Xu9nLPXf3bnzer5k7e85znnueb87efPbsc+85m6pCktSOJ0y6AEnS0jL4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY1ZMekCBjn00ENr1apVky5DkvYb119//QNVNTVM32UZ/KtWrWLr1q2TLkOS9htJvjZsX6d6JKkxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMcvyyt0+Vp1zxUTG3X7uqycyriQ9Xp7xS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjVnwXj1JLgJOAXZV1fO7tsuAo7suBwPfrKrjBjx3O/At4AfAnqqaHlPdkqQRDXOTtouB84AP722oqt/Yu5zkb4GH9/H8E6rqgVELlCSN14LBX1XXJFk1aFuSAK8Hfnm8ZUmSFkvfOf5fAu6vqjvn2V7A1UmuT7Kh51iSpDHoez/+9cCl+9j+sqrameSpwJYkt1fVNYM6dj8YNgCsXLmyZ1mSpPmMfMafZAXwa8Bl8/Wpqp3d113A5cCaffTdWFXTVTU9NTU1almSpAX0mep5JXB7Ve0YtDHJgUkO2rsMnAjc0mM8SdIYLBj8SS4FvgAcnWRHkrO6TeuYM82T5OlJNnerhwHXJrkR+BJwRVVdNb7SJUmjGOZTPevnaT9zQNvXgbXd8j3AsT3rkySNmVfuSlJjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0Z5o+tX5RkV5JbZrW9K8nOJNu6x9p5nntSkjuS3JXknHEWLkkazTBn/BcDJw1of39VHdc9Ns/dmOQA4HzgZOAYYH2SY/oUK0nqb8Hgr6prgIdG2Pca4K6quqeqvg98DDhthP1Iksaozxz/W5Lc1E0FHTJg+xHAvbPWd3RtAyXZkGRrkq27d+/uUZYkaV9GDf4PAM8GjgPuA/62byFVtbGqpqtqempqqu/uJEnzGCn4q+r+qvpBVf0Q+CAz0zpz7QSOmrV+ZNcmSZqgkYI/yeGzVl8L3DKg25eB1UmemeRJwDpg0yjjSZLGZ8VCHZJcChwPHJpkB/BO4PgkxwEFbAfe2PV9OvChqlpbVXuSvAX4FHAAcFFV3boo/wpJ0tAWDP6qWj+g+cJ5+n4dWDtrfTPwmI96SpImxyt3JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1ZsHgT3JRkl1JbpnV9t4ktye5KcnlSQ6e57nbk9ycZFuSreMsXJI0mmHO+C8GTprTtgV4flW9APgq8Of7eP4JVXVcVU2PVqIkaZwWDP6qugZ4aE7b1VW1p1u9DjhyEWqTJC2Ccczx/x5w5TzbCrg6yfVJNoxhLElSTyv6PDnJ24E9wEfm6fKyqtqZ5KnAliS3d79BDNrXBmADwMqVK/uUJUnah5HP+JOcCZwC/GZV1aA+VbWz+7oLuBxYM9/+qmpjVU1X1fTU1NSoZUmSFjBS8Cc5CfhT4NSq+s48fQ5MctDeZeBE4JZBfSVJS2eYj3NeCnwBODrJjiRnAecBBzEzfbMtyQVd36cn2dw99TDg2iQ3Al8CrqiqqxblXyFJGtqCc/xVtX5A84Xz9P06sLZbvgc4tld1kqSx88pdSWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNGSr4k1yUZFeSW2a1PTnJliR3dl8Pmee5Z3R97kxyxrgKlySNZtgz/ouBk+a0nQN8uqpWA5/u1n9EkicD7wReDKwB3jnfDwhJ0tIYKvir6hrgoTnNpwGXdMuXAK8Z8NRfAbZU1UNV9Q1gC4/9ASJJWkJ95vgPq6r7uuX/Bg4b0OcI4N5Z6zu6NknShIzlzd2qKqD67CPJhiRbk2zdvXv3OMqSJA3QJ/jvT3I4QPd114A+O4GjZq0f2bU9RlVtrKrpqpqemprqUZYkaV/6BP8mYO+ndM4A/nVAn08BJyY5pHtT98SuTZI0IcN+nPNS4AvA0Ul2JDkLOBd4VZI7gVd26ySZTvIhgKp6CHgP8OXu8e6uTZI0ISuG6VRV6+fZ9IoBfbcCb5i1fhFw0UjVSZLGzit3JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1ZuTgT3J0km2zHo8keeucPscneXhWn3f0L1mS1MdQf3N3kKq6AzgOIMkBwE7g8gFdP19Vp4w6jiRpvMY11fMK4O6q+tqY9idJWiTjCv51wKXzbHtpkhuTXJnkeWMaT5I0ot7Bn+RJwKnAxwdsvgF4RlUdC/wD8Ml97GdDkq1Jtu7evbtvWZKkeYzjjP9k4Iaqun/uhqp6pKoe7ZY3A09McuignVTVxqqarqrpqampMZQlSRpkHMG/nnmmeZI8LUm65TXdeA+OYUxJ0ohG/lQPQJIDgVcBb5zV9iaAqroAOB14c5I9wHeBdVVVfcaUJPXTK/ir6tvAU+a0XTBr+TzgvD5jSJLGyyt3JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMb0Dv4k25PcnGRbkq0DtifJ3ye5K8lNSV7Ud0xJ0uh6/bH1WU6oqgfm2XYysLp7vBj4QPdVkjQBSzHVcxrw4ZpxHXBwksOXYFxJ0gDjCP4Crk5yfZINA7YfAdw7a31H1/YjkmxIsjXJ1t27d4+hLEnSIOMI/pdV1YuYmdI5O8nLR9lJVW2squmqmp6amhpDWZKkQXoHf1Xt7L7uAi4H1szpshM4atb6kV2bJGkCegV/kgOTHLR3GTgRuGVOt03A73Sf7nkJ8HBV3ddnXEnS6Pp+qucw4PIke/f10aq6KsmbAKrqAmAzsBa4C/gO8Ls9x5Qk9dAr+KvqHuDYAe0XzFou4Ow+40iSxscrdyWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9Jjen7F7ikiVh1zhUTGXf7ua+eyLjSOHnGL0mNGTn4kxyV5LNJbktya5I/HtDn+CQPJ9nWPd7Rr1xJUl99pnr2AG+rqhuSHARcn2RLVd02p9/nq+qUHuNIksZo5DP+qrqvqm7olr8FfAU4YlyFSZIWx1jm+JOsAl4IfHHA5pcmuTHJlUmeN47xJEmj6/2pniQ/Dfwz8NaqemTO5huAZ1TVo0nWAp8EVs+znw3ABoCVK1f2LUuSNI9eZ/xJnshM6H+kqv5l7vaqeqSqHu2WNwNPTHLooH1V1caqmq6q6ampqT5lSZL2oc+negJcCHylqt43T5+ndf1IsqYb78FRx5Qk9ddnqucXgd8Gbk6yrWv7C2AlQFVdAJwOvDnJHuC7wLqqqh5jSpJ6Gjn4q+paIAv0OQ84b9QxJEnj5y0bJGmOH/dbgnjLBklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mN8ZYNYzKpS7y1tH7cL+VfTvw/tXg845ekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mN6RX8SU5KckeSu5KcM2D7TyS5rNv+xSSr+ownSepv5OBPcgBwPnAycAywPskxc7qdBXyjqn4OeD/w16OOJ0kajz5n/GuAu6rqnqr6PvAx4LQ5fU4DLumWPwG8Isk+/0C7JGlx9Qn+I4B7Z63v6NoG9qmqPcDDwFN6jClJ6mnZ3LIhyQZgQ7f6aJI79tH9UOCBxa9qbKx3cf3Y15vJTpLuT8d3f6oV5tTb8/v8jGE79gn+ncBRs9aP7NoG9dmRZAXws8CDg3ZWVRuBjcMMnGRrVU0/7oonxHoXl/Uurv2p3v2pVphcvX2mer4MrE7yzCRPAtYBm+b02QSc0S2fDnymqqrHmJKknkY+46+qPUneAnwKOAC4qKpuTfJuYGtVbQIuBP4xyV3AQ8z8cJAkTVCvOf6q2gxsntP2jlnL3wN+vc8Y8xhqSmgZsd7FZb2La3+qd3+qFSZUb5x5kaS2eMsGSWrMsg7+/e2WEEPU+/IkNyTZk+T0SdQ4p56F6v2TJLcluSnJp5MM/XGxxTBEvW9KcnOSbUmuHXAl+ZJZqNZZ/V6XpJJM9JMoQxzbM5Ps7o7ttiRvmESds+pZ8PgmeX33+r01yUeXusY5tSx0fN8/69h+Nck3F7WgqlqWD2beML4beBbwJOBG4Jg5ff4AuKBbXgdctszrXQW8APgwcPp+cHxPAH6qW37zfnB8f2bW8qnAVcu11q7fQcA1wHXA9DI/tmcC502qxhHqXQ38J3BIt/7U5VzvnP5/yMyHZRatpuV8xr+/3RJiwXqrantV3QT8cBIFzjFMvZ+tqu90q9cxc63GpAxT7yOzVg8EJvUG1jCvXYD3MHP/qu8tZXEDDFvvcjFMvb8PnF9V3wCoql1LXONsj/f4rgcuXcyClnPw72+3hBim3uXk8dZ7FnDlola0b0PVm+TsJHcDfwP80RLVNteCtSZ5EXBUVS2Hvyg+7Gvhdd203yeSHDVg+1IZpt7nAM9J8h9Jrkty0pJV91hD/1/rplOfCXxmMQtazsGvZSLJbwHTwHsnXctCqur8qno28GfAX066nkGSPAF4H/C2SdfyOPwbsKqqXgBs4f9/016uVjAz3XM8M2fQH0xy8EQrGs464BNV9YPFHGQ5B//juSUEC90SYgkMU+9yMlS9SV4JvB04tar+Z4lqG+TxHt+PAa9Z1Irmt1CtBwHPBz6XZDvwEmDTBN/gXfDYVtWDs77/HwJ+folqG2SY18IOYFNV/W9V/RfwVWZ+EEzC43ntrmORp3mAZf3m7grgHmZ+7dn7hsjz5vQ5mx99c/eflnO9s/pezOTf3B3m+L6QmTelVu8nr4fVs5Z/lZkryJdlrXP6f47Jvrk7zLE9fNbya4Hrlnm9JwGXdMuHMjPV8pTlWm/X77nAdrrrqxa1pkl984Y8YGuZ+Ul9N/D2ru3dzJx9Avwk8HHgLuBLwLOWeb2/wMyZyLeZ+c3k1mVe778D9wPbusemZV7v3wG3drV+dl9hO+la5/SdaPAPeWz/qju2N3bH9rnLvN4wM512G3AzsG4519utvws4dynq8cpdSWrMcp7jlyQtAoNfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TG/B/p/3ckdOXuOQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist((cg<hthresh).mean(axis=0))\n",
    "good_celltypes = (cg<hthresh).mean(axis=0)<0.2\n",
    "good_celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = df_mlevel.columns\n",
    "celltypes = [re.sub(\"(Round2\\.|_mCG)\",\"\",c) for c in celltypes]\n",
    "df_mlevel.columns = celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell type names\n",
    "df_cellnames = pd.read_csv('/cndd/Public_Datasets/CEMBA/BICCN_minibrain_data/data_freeze/supp_info/clusters_final/cluster_annotation_scf_round2.tsv',\n",
    "                       sep='\\t', index_col='cluster')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load ATAC data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac = pd.read_csv(datadir+'enhancers_reptile_atac_norm.tsv',sep='\\t')\n",
    "atac=atac.set_index('enhancer_idx',drop=False)\n",
    "atac_enh_list = pd.read_csv(datadir+'enhancers_reptile.bed',sep='\\t',header=None,names=['chr','start','end'])\n",
    "atac = atac.join(atac_enh_list)\n",
    "atac['enh_pos'] = atac['chr'].astype(str)+'_'+atac['start'].astype(str)+'_'+atac['end'].astype(str)\n",
    "atac = atac.loc[~(atac['chr']==\"chrY\")]\n",
    "df_atac = atac.set_index('enh_pos',drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure the ATAC and mCG data are both in the same column order\n",
    "df_atac = df_atac[celltypes]\n",
    "df_mlevel = df_mlevel[celltypes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = df_mlevel.columns\n",
    "celltypes = [re.sub(\"(Round2\\.|_mCG)\",\"\",c) for c in celltypes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "kdict=np.load(datadir+'kmers_6mers/kmer_dict.npy', allow_pickle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the kmer data per enhancer\n",
    "kmer_counts=np.load(datadir+'kmers_6mers/count.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_enh = pd.read_csv(datadir+'enhancers_reptile.bed', sep='\\t', \n",
    "                       header=None, names=['chr','start','end'])\n",
    "kmer_enh['enh_pos'] = ['_'.join([str(i) for i in row]) for ind,row in kmer_enh.iterrows()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_enh['enh_num'] = np.arange(kmer_enh.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(302106, 2080)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_enh = kmer_enh.set_index('enh_pos', drop=False)\n",
    "kmer_counts.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare the data and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((302096, 29), (142890, 29))"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure all data frames have the cell types in the same order\n",
    "enh_posu = genes2enhu['enh_pos'].unique()\n",
    "df_mlevelu = df_mlevel.loc[enh_posu]\n",
    "df_atacu = df_atac.loc[enh_posu]\n",
    "df_mlevel.shape, df_mlevelu.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Determine the enhancer number for each gene-enhancer pair\n",
    "genes2enhu = genes2enh.loc[rna.index]  # Note that this is larger than df_mlevelu\n",
    "genes2enhu['enh_num'] = kmer_enh.loc[genes2enhu['enh_pos'],'enh_num'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_countsu = kmer_counts[genes2enhu['enh_num'],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((142890, 21), (142890, 21))"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop the bad cell types\n",
    "bad_celltypes = cg.columns[((cg<hthresh).mean(axis=0)>0.2)]\n",
    "bad_celltypes = [i.replace('Round2.','').replace('_mCG','') for i in bad_celltypes]\n",
    "rnau = rna.drop(bad_celltypes,axis=1).drop(['Unnamed: 0'],axis=1)\n",
    "df_mlevelu = df_mlevelu.loc[:,rnau.columns]\n",
    "df_atacu = df_atacu.loc[:,rnau.columns]\n",
    "assert( (rnau.columns==df_mlevelu.columns).all() )\n",
    "assert( (rnau.columns==df_atacu.columns).all() )\n",
    "df_mlevelu.shape, df_atacu.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RegressData/Regress_data_9868genes_21cells_genes2enhu.pkl\n",
      "RegressData/Regress_data_9868genes_21cells_rnau.pkl\n",
      "RegressData/Regress_data_9868genes_21cells_df_mlevelu.pkl\n",
      "RegressData/Regress_data_9868genes_21cells_df_atacu.pkl\n",
      "RegressData/Regress_data_9868genes_21cells_genes.pkl\n"
     ]
    }
   ],
   "source": [
    "# Save the data needed for training\n",
    "fn_save_prefix = 'RegressData/Regress_data_%dgenes_%dcells_' % (rnau.shape[0], rnau.shape[1])\n",
    "save_vars = ['genes2enhu', 'rnau', 'df_mlevelu', 'df_atacu','genes']\n",
    "for var in save_vars:\n",
    "    fn_save = fn_save_prefix+var+'.pkl'\n",
    "    eval(var).to_pickle(fn_save)\n",
    "    print(fn_save)\n",
    "np.savez_compressed('RegressData/Regress_Feb16_data_kmer_countsu.npz', kmer_countsu=kmer_countsu)\n",
    "print(fn_save_prefix +'kmer_countsu.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Impute the NaN values in mlevel\n",
    "df_mlevelu_mean = np.outer(df_mlevelu.median(axis=1).values, df_mlevelu.median(axis=0).values)\n",
    "df_mlevelu_mean = pd.DataFrame(df_mlevelu_mean, index=df_mlevelu.index, columns=df_mlevelu.columns)\n",
    "df_mlevelu_mean = df_mlevelu_mean.fillna(df_mlevelu.median())\n",
    "df_mlevelu = df_mlevelu.fillna(df_mlevelu_mean)\n",
    "assert(not df_mlevelu.isna().any(axis=None))\n",
    "assert(not df_atacu.isna().any(axis=None))\n",
    "ensids = rnau.index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out bad enhancers\n",
    "bad_enh = df_mlevelu.loc[genes2enhu.loc[:,'enh_pos']].isna().any(axis=1)\n",
    "bad_enh = bad_enh | df_atacu.loc[genes2enhu.loc[:,'enh_pos']].isna().any(axis=1)\n",
    "good_enh = [not i for i in bad_enh]\n",
    "genes2enhu = genes2enhu[good_enh]\n",
    "genes2enhu['enh_num2'] = np.arange(genes2enhu.shape[0])\n",
    "kmer_countsu = kmer_countsu[good_enh]\n",
    "K = kmer_countsu.shape[1] # Number of K-mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather all the mC and ATAC data into a numpy array for quick indexing\n",
    "# %%timeit\n",
    "ensids = rnau.index.values\n",
    "max_nenh = genes2enhu.groupby('ensid')['gene_chr'].count().max()\n",
    "Ne = ensids.shape[0]\n",
    "Nc = df_mlevelu.shape[1]\n",
    "rna_lookup = rnau.loc[ensids].to_numpy()\n",
    "ml_lookup = np.zeros((Ne,max_nenh,Nc))\n",
    "atac_lookup = np.zeros((Ne,max_nenh,Nc))\n",
    "kmeri_lookup = [];\n",
    "for i,ens in enumerate(ensids):\n",
    "    curr_enh = genes2enhu.loc[ens,['enh_pos','enh_num2']]\n",
    "    mlu = df_mlevelu.loc[curr_enh['enh_pos']].to_numpy()\n",
    "    atacu = df_atacu.loc[curr_enh['enh_pos']]\n",
    "    ml_lookup[i,:mlu.shape[0],:] = mlu\n",
    "    atac_lookup[i,:atacu.shape[0],:] = atacu\n",
    "    kmeri_lookup.append(curr_enh['enh_num2'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savez_compressed(fn_save_prefix+'AllLookups.npz',\n",
    "                   ml_lookup=ml_lookup,\n",
    "                   atac_lookup=atac_lookup,\n",
    "                   rna_lookup=rna_lookup,\n",
    "                   kmeri_lookup=kmeri_lookup)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
